core/network/src/protocol/light_client_handler.rs

```diff
@@ -0,0 +1,1667 @@
+			status: PeerStatus::Idle,
+		};
+
+		self.peers.insert(peer, info);
+	}
+
+	fn inject_disconnected(&mut self, peer: &PeerId, _: ConnectedPoint) {
+		trace!("peer {} disconnected", peer);
+		self.remove_peer(peer)
+	}
+
+	fn inject_node_event(&mut self, peer: PeerId, event: Event<T>) {
+		match event {
+			// An incoming request from remote has been received.
+			Event::Request(request, stream) => {
+				let result = match &request.request {
+					Some(api::v1::light::request::Request::RemoteCallRequest(r)) =>
+						self.on_remote_call_request(&peer, request.id, r),
+					Some(api::v1::light::request::Request::RemoteReadRequest(r)) =>
+						self.on_remote_read_request(&peer, request.id, r),
+					Some(api::v1::light::request::Request::RemoteHeaderRequest(r)) =>
+						self.on_remote_header_request(&peer, request.id, r),
+					Some(api::v1::light::request::Request::RemoteReadChildRequest(r)) =>
+						self.on_remote_read_child_request(&peer, request.id, r),
+					Some(api::v1::light::request::Request::RemoteChangesRequest(r)) =>
+						self.on_remote_changes_request(&peer, request.id, r),
+					None => {
+						debug!("ignoring request {} without request data from peer {}", request.id, peer);
+						return
+					}
+				};
+				match result {
+					Ok(response) => {
+						trace!("enqueueing response {} for peer {}", response.id, peer);
+						let mut data = Vec::new();
+						if let Err(e) = response.encode(&mut data) {
+							debug!("error encoding response {} for peer {}: {}", response.id, peer, e)
+						} else {
+							self.responses.push_back(write_one(stream, data))
+						}
+					}
+					Err(e) => debug!("error handling request {} from peer {}: {}", request.id, peer, e)
+				}
+			}
+			// A response to one of our own requests has been received.
+			Event::Response(response) => {
+				let id = response.id;
+				if let Some(request) = self.outstanding.remove(&id) {
+					// We first just check if the response originates from the expected peer.
+					if request.peer != peer {
+						debug!("was expecting response {} from {} instead of {}", id, request.peer, peer);
+						self.outstanding.insert(id, request);
+						self.peerset.report_peer(peer.clone(), i32::min_value());
+						self.remove_peer(&peer);
+						return
+					}
+
+					if let Some(info) = self.peers.get_mut(&peer) {
+						if info.status != PeerStatus::BusyWith(id) {
+							// If we get here, something is wrong with our internal handling of peer
+							// status information. At any time, a single peer processes at most one
+							// request from us and its status should contain the request ID we are
+							// expecting a response for. If a peer would send us a response with a
+							// random ID, we should not have an entry for it with this peer ID in
+							// our `outstanding` map, so a malicious peer should not be able to get
+							// us here. It is our own fault and must be fixed!
+							panic!("unexpected peer status {:?} for {}", info.status, peer);
```

Comment:
I'd go for calling `log::error!` and returning instead.

Reply:
I do not think it makes sense to continue past that point as the internal state is inconsistent. As mentioned in the comment, this should never be reached. If it is however, the programme as written is wrong and needs to be fixed.

It would be nice if the `NetworkBehaviour` trait provides a way to signal an error condition to the swarm which would then stop polling it.

Reply:
@twittner our codebase policy with panickers is to prove that they won't happen or remove them - is it not possible to recover again to a valid protocol state? mild programmer error should not take down a decentralized network

Reply:
If you prefer to keep the behaviour going no matter what, we could maybe attempt to reset all internal state.

Reply:
Is there a way to kill the node gracefully then?

Reply:
No that I am aware of. The `NetworkBehaviour` trait offers no such mechanism at the moment AFAIK. This would need to be addressed on libp2p-side first. CC @tomaka.

Reply:
I have opened https://github.com/libp2p/rust-libp2p/issues/1256 to discuss this further.

Reply:
It doesn't make sense to have in libp2p a generic mechanism to say "hey there's a bug in the code, please do something".
The behaviour can however emit an event that can correspond to anything we want.